lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,496 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// +build !solaris,!darwin solaris,cgo darwin,cgo
+
+package fswatcher
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/ignore"
+	"github.com/zillode/notify"
+)
+
+type fsEventType int
+
+const (
+	nonRemove fsEventType = 1
+	remove                = 2
+	mixed                 = 3
+)
+
+type fsEvent struct {
+	path         string
+	firstModTime time.Time
+	lastModTime  time.Time
+	eventType    fsEventType
+}
+
+type fsEventsBatch map[string]*fsEvent
+
+type eventDir struct {
+	path      string
+	parentDir *eventDir
+	events    fsEventsBatch
+	dirs      map[string]*eventDir
+}
+
+func newEventDir(path string, parentDir *eventDir) *eventDir {
+	return &eventDir{
+		path:      path,
+		parentDir: parentDir,
+		events:    make(fsEventsBatch),
+		dirs:      make(map[string]*eventDir),
+	}
+}
+
+type fsWatcher struct {
+	folderPath      string
+	notifyModelChan chan []string
+	// All detected and to be scanned events are stored in a tree like
+	// structure mimicking folders to keep count of events per directory.
+	rootEventDir *eventDir
+	fsEventChan  chan notify.EventInfo
+	// time interval to search for events to be passed to syncthing-core
+	notifyDelay time.Duration
+	// time after which an active event is passed to syncthing-core
+	notifyTimeout         time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	resetNotifyTimerChan  chan time.Duration
+	inProgress            map[string]struct{}
+	description           string
+	ignores               *ignore.Matcher
+	ignoresUpdate         chan *ignore.Matcher
+	stop                  chan struct{}
+	ignorePerms           bool
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	FsWatchChan() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+}
+
+// Not meant to be changed, but must be changeable for tests
+var (
+	maxFiles       = 512
+	maxFilesPerDir = 128
+)
+
+func NewFsWatcher(cfg config.FolderConfiguration, ignores *ignore.Matcher) (Service, error) {
+	fsWatcher := &fsWatcher{
+		folderPath:            filepath.Clean(cfg.Path()),
+		notifyModelChan:       make(chan []string),
+		rootEventDir:          newEventDir(".", nil),
+		fsEventChan:           make(chan notify.EventInfo, maxFiles),
+		notifyDelay:           time.Duration(cfg.NotifyDelayS) * time.Second,
+		notifyTimeout:         notifyTimeout(cfg.NotifyDelayS),
+		notifyTimerNeedsReset: false,
+		inProgress:            make(map[string]struct{}),
+		description:           cfg.Description(),
+		ignores:               ignores,
+		ignoresUpdate:         make(chan *ignore.Matcher),
+		resetNotifyTimerChan:  make(chan time.Duration),
+		stop:                  make(chan struct{}),
+		ignorePerms:           cfg.IgnorePerms,
+	}
```

Comment:
Should be doable. On that account I have a question about `(f FolderConfiguration) Path()`: The notify package returns absolute paths in the filepath format (i.e. as `filepath.Clean`) and the filepath package is used to check/manipulate paths in fswatcher. In `cleanedPath()` in folderconfiguration.go there is some windows specific stuff going on that I do not understand. Previously I just set `fsWatcher.filepath()` to `filepath.Clean(cfg.Path())` hoping for a sane result. Is this necessary or can I just assume `cfg.Path()` will work for stuff like checking whether an event lies within the folder root directory?